imports/composite_redolog_manager.go

```diff
@@ -0,0 +1,330 @@
+	redoFilePersisited int64
+	offsetPersisted    uint32
+	// flag to indicate the ingestion from kafka is dead
+	done bool
+	// message counts
+	msgRecovered int64
+	msgReceived  int64
+	msgError     int64
+}
+
+// NewCompositeRedologManager create CompositeRedologManager oibject
+func NewCompositeRedologManager(table string, shard int, factory *RedologManagerFactory, tableConfig *metaCom.TableConfig, storeFunc BatchStoreFunc) *CompositeRedologManager {
+	var fileRedoLogManager *FileRedologManager
+	if factory == nil || factory.enableLocalRedoLog {
+		fileRedoLogManager = NewFileRedoLogManager(int64(tableConfig.RedoLogRotationInterval), int64(tableConfig.MaxRedoLogFileSize), factory.diskStore, table, shard)
+	}
+	var kafkaReader *kafkaPartitionReader
+
+	if factory != nil && factory.consumer != nil {
+		commitFunc := func(offset int64) error {
+			return factory.metaStore.UpdateIngestionCommitOffset(table, shard, offset)
+		}
+		checkPointFunc := func(offset int64) error {
+			return factory.metaStore.UpdateIngestionCheckpointOffset(table, shard, offset)
+		}
+		kafkaReader = NewKafkaPartitionReader(factory.namespace, table, shard, factory.consumer, commitFunc, checkPointFunc)
+	}
+	manager := &CompositeRedologManager{
+		Table:              table,
+		Shard:              shard,
+		factory:            factory,
+		fileRedoLogManager: fileRedoLogManager,
+		kafkaReader:        kafkaReader,
+		readyChan:          make(chan bool, 1),
+		storeFunc:          storeFunc,
+	}
+
+	return manager
+}
+
+// convenient function for backward test
+func (s *CompositeRedologManager) GetLocalFileRedologManager() *FileRedologManager {
+	return s.fileRedoLogManager
+}
+
+// convenient function for backward test
+func (s *CompositeRedologManager) GetKafkaReader() *kafkaPartitionReader {
+	return s.kafkaReader
+}
+
+// Start start the recovery and data ingestion routine
+func (s *CompositeRedologManager) Start(redoLogFilePersisted int64, offsetPersisted uint32) error {
+	s.redoFilePersisited = redoLogFilePersisted
+	s.offsetPersisted = offsetPersisted
+
+	go func() {
+		// first read from local redolog file for recovery if it is enabled
+		if s.fileRedoLogManager != nil {
+			utils.GetLogger().With("action", "ingestion", "table", s.Table, "shard", s.Shard).
+				Info("Start recovery from local redolog")
+			fn := s.fileRedoLogManager.NextUpsertBatch()
+			for {
+				batch, redoLogFile, batchOffset := fn()
+				if batch == nil {
+					break
+				} else {
+					if err := s.applyUpsertBatchOnly(batch, redoLogFile, batchOffset); err != nil {
+						utils.GetLogger().With("err", err).Panic("Failed to apply upsert batch during recovery")
+					}
+				}
+			}
+			s.setRecoveryDone()
+			utils.GetLogger().With("action", "ingestion", "table", s.Table, "shard", s.Shard,
+				"batchRecovered", s.msgRecovered).Info("Finished recovery from local redolog")
+		}
+		// next read from kafka if kafka input is enabled
+		if s.kafkaReader != nil {
+			var offsetFrom, offsetTo int64
```

Comment:
the logic is similar to local redolog , try abstract the common part and reuse the function